### PR TITLE
@@isConcatSpreadable for Tuples

### PR DESCRIPTION
**Pull request by rickbutton**

### CODE REVIEW

**Comment by littledan** [comment]
IMO we should follow Arrays and update the [IsConcatSpreadable](https://tc39.es/ecma262/#sec-isconcatspreadable) algorithm directly instead of defining this property, while still respecting that property if someone does define it on `Tuple.prototype`.

**Comment by ljharb** [comment]
good call

File: spec/abstract-operations.html [hunk 1]
```diff
@@ -744,4 +744,17 @@ <h1>AddValueToTupleSequenceList ( _sequence_, _value_ )</h1>
...
   </emu-clause>
+
+  <emu-clause id="sec-isconcatspreadable" aoid="IsConcatSpreadable">
+      <h1>Runtime Semantics: IsConcatSpreadable ( _O_ )</h1>
+      <p>The abstract operation IsConcatSpreadable takes argument _O_. It performs the following steps when called:</p>
+      <emu-alg>
+        1. If Type(_O_) is not Object, return *false*.
+        1. Let _spreadable_ be ? Get(_O_, @@isConcatSpreadable).
+        1. If _spreadable_ is not *undefined*, return ! ToBoolean(_spreadable_).
+        1. <ins>If _O_ has a [[TupleData]] internal slot, return *true*</ins>
```
Comment (littledan) [inline spec/abstract-operations.html:755]: If Type(_O_) is Tuple or if Type(_O_) is Object and _O_ has a [[TupleData]] internal slot, return *true*

File: spec/abstract-operations.html [hunk 2]
```diff
@@ -744,4 +744,17 @@ <h1>AddValueToTupleSequenceList ( _sequence_, _value_ )</h1>
       </emu-alg>
     </emu-clause>
   </emu-clause>
+
+  <emu-clause id="sec-isconcatspreadable" aoid="IsConcatSpreadable">
+      <h1>Runtime Semantics: IsConcatSpreadable ( _O_ )</h1>
+      <p>The abstract operation IsConcatSpreadable takes argument _O_. It performs the following steps when called:</p>
+      <emu-alg>
+        1. If Type(_O_) is not Object, return *false*.
```
Comment (littledan) [inline spec/abstract-operations.html:752]: Oops, we already returned false up here! "is not Object or Tuple", probably this should say, with a GetV below.

**Comment by rickbutton** [comment]
@littledan apologies, I iterated on this PR a bunch of times in the middle of your review. Making tuples -always- concat spreadable was a mistake, I moved the check to after the symbol lookup. I think it would be weird if Tuples were the only thing that was forced into concat-spreadable with no opt opt.

File: spec/abstract-operations.html [hunk 3]
```diff
@@ -744,4 +744,17 @@ <h1>AddValueToTupleSequenceList ( _sequence_, _value_ )</h1>
...
   </emu-clause>
+
+  <emu-clause id="sec-isconcatspreadable" aoid="IsConcatSpreadable">
+      <h1>Runtime Semantics: IsConcatSpreadable ( _O_ )</h1>
+      <p>The abstract operation IsConcatSpreadable takes argument _O_. It performs the following steps when called:</p>
+      <emu-alg>
+        1. If Type(_O_) is not Object or Tuple, return *false*
+        1. Let _spreadable_ be ? Get<ins>V</ins>(_O_, @@isConcatSpreadable).
+        1. If _spreadable_ is not *undefined*, return ! ToBoolean(_spreadable_).
+        1. <ins>If Type(_O_) is Tuple or Type(_O_) is Object and _O_ has a [[TupleData]] internal slot, return *true*</ins>
```
Comment (ljharb) [inline spec/abstract-operations.html:755]: is there no internal `isTuple` AO for this?

Comment (rickbutton) [reply spec/abstract-operations.html:755]: not currently, we are manually checking for `TupleData` in the few places we need to. We could certainly add one.

Comment (ljharb) [reply spec/abstract-operations.html:755]: I'd prefer that for spec clarity :-)

Comment (rickbutton) [reply spec/abstract-operations.html]: sure thing, just added, also switched `Tuple.isTuple` over to using it.

**Pull request merged**